src/Symfony/Component/Messenger/Command/FailedMessagesRetryCommand.php

```diff
@@ -0,0 +1,181 @@
+
+    <info>php %command.full_name% {id}</info>
+
+Or pass multiple ids at once to process multiple messages:
+
+<info>php %command.full_name% {id1} {id2} {id3}</info>
+
+EOF
+            )
+        ;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $io = new SymfonyStyle($input, $output instanceof ConsoleOutputInterface ? $output->getErrorOutput() : $output);
+        $io->comment('Quit this command with CONTROL-C.');
+        if (!$output->isVeryVerbose()) {
+            $io->comment('Re-run the command with a -vv option to see logs about consumed messages.');
+        }
+
+        $receiver = $this->getReceiver();
+        if ($receiver instanceof MessageCountAwareInterface) {
+            if (1 === $receiver->getMessageCount()) {
+                $io->writeln('There is <comment>1</comment> message waiting in the failure transport.');
+            } else {
+                $io->writeln(sprintf('There are <comment>%d</comment> messages waiting in the failure transport.', $receiver->getMessageCount()));
+            }
+        }
+
+        $io->writeln(sprintf('To retry all the messages, run <comment>messenger:consume %s</comment>', $this->getReceiverName()));
+
+        $shouldForce = $input->getOption('force');
+        $ids = $input->getArgument('id');
+        if (0 === \count($ids)) {
+            if (!$input->isInteractive()) {
+                throw new RuntimeException('Message id must be passed when in non-interactive mode.');
+            }
+
+            return $this->runInteractive($io, $shouldForce);
+        }
+
+        if (!$receiver instanceof ListableReceiverInterface) {
+            throw new RuntimeException(sprintf('The "%s" receiver does not support retrying messages by id.', $this->getReceiverName()));
+        }
+
+        foreach ($ids as $id) {
+            $envelope = $receiver->find($id);
+            if (null === $envelope) {
+                throw new RuntimeException(sprintf('The message "%s" was not found.', $id));
+            }
+
+            $singleReceiver = new SingleMessageReceiver($receiver, $envelope);
+            $this->runWorker($singleReceiver, $io, $shouldForce);
+        }
+        $io->success('All done!');
+    }
+
+    private function runInteractive(SymfonyStyle $io, bool $shouldForce)
+    {
+        $this->runWorker($this->getReceiver(), $io, $shouldForce);
+
+        $io->success('All failed messages have been handled or removed!');
+    }
+
+    private function runWorker(ReceiverInterface $receiver, SymfonyStyle $io, bool $shouldForce)
+    {
+        $listener = function (WorkerMessageReceivedEvent $messageReceivedEvent) use ($io, $receiver, $shouldForce) {
+            $envelope = $messageReceivedEvent->getEnvelope();
+
+            $this->displaySingleMessage($envelope, $io);
+
+            $shouldHandle = $shouldForce || $io->confirm('Do you want to retry (yes) or delete this message (no)?');
```

Comment:
I'm facing a weird behavior. 
- I have 1 message waiting in the failure transport 
- I run this command
- I use `CONTROL-C` at time to answer this confirmation question (interrupt before answering) 

Then the failure transport is left in a weird state: 
- `messenger:failed:show` says there are 0 waiting messages, same for `messenger:failed:retry`, *but* not for `messenger:failed:retry MESSAGE_ID `which does find the failed message although it tells there are 0 waiting messages

Here is a screenshot: https://prnt.sc/ni19u1. I believe the message gets kind of semi-handled since the receiver still `get()`s it but I can't explain why `getMessageCount()` returns `0`.
This bug seems to only occur when I have only one failed message waiting, and I'm using an sqlite doctrine failure transport. Please tell me if you can reproduce or explain :)

Reply:
Great catch. Less robust transports (e.g. Doctrine) have a problem that more robust transports (e.g. amqp) do not have: when you "fetch" a message via the standard `ReceiverInterface::get()`, less-robust transports will automatically mark that as "acknowledged" so that no other worker will use it. If handling it is interrupted, it remains "acked". That's not a *real* problem, because Doctrine has a built-in mechanism to make those message available again after some time (1 hour by default).

But, it *does* cause this weird behavior your described. The fix (that I've just implemented) is:
A) if the failure transport implements `ListableReceiverInterface`, fetch messages one-by-one with `all()` and handle them. This is always a safe operation, so we use it when possible.
B) If not, use the normal mechanism. AMQP will use this normal mechanism, but it also doesn't have the problem you described above.